### PR TITLE
GMS-991: Add `safeBurn` and `safeBurnBatch`.

### PR DESCRIPTION
**Pull request by dlevy47**

### CODE REVIEW

File: contracts/token/erc721/abstract/ImmutableERC721Base.sol [hunk 1]
```diff
@@ -216,6 +222,26 @@ abstract contract ImmutableERC721Base is
...
+        address currentOwner = ownerOf(tokenId);
+        if (currentOwner != owner) {
+            revert IImmutableERC721MismatchedTokenOwner(tokenId, currentOwner);
+        }
+
+        burn(tokenId);
+    }
+
+    /// @dev Burn a batch of tokens, checking the owner of each token first.
+    function safeBurnBatch(IDBurn[] memory burns) public virtual {
```
Comment (jasonzwli) [inline contracts/token/erc721/abstract/ImmutableERC721Base.sol:236]: can we expose this on the preset instead?

File: test/token/erc721/ImmutableERC721HybridPermissionedMintable.test.ts [hunk 2]
```diff
@@ -163,6 +163,72 @@ describe("ImmutableERC721", function () {
       );
     });
 
+    it("Should not allow owner or approved to burn a token when specifying the incorrect owner", async function() {
+      await expect(
+        erc721.connect(user).safeBurn(owner.address, 5)
+      ).to.be.revertedWith('IImmutableERC721MismatchedTokenOwner');
```
Comment (jasonzwli) [inline test/token/erc721/ImmutableERC721HybridPermissionedMintable.test.ts:169]: lets add the `withArgs` after the `revertedWith`

File: contracts/token/erc721/abstract/ImmutableERC721Base.sol (hunk 1 above)
Comment (dlevy47) [reply contracts/token/erc721/abstract/ImmutableERC721Base.sol]: Done.

File: test/token/erc721/ImmutableERC721HybridPermissionedMintable.test.ts (hunk 2 above)
Comment (dlevy47) [reply test/token/erc721/ImmutableERC721HybridPermissionedMintable.test.ts]: Done.

**Review by jasonzwli — APPROVED** [review]

**Review by ipekt — APPROVED** [review]

**Pull request merged**